org-code-javabuilder/app/src/main/java/org/code/javabuilder/JavaBuildController.java

```diff
@@ -21,9 +38,122 @@
+      boolean success = task.call();
+
+      // diagnostics will include any compiler errors
+      for (Diagnostic<? extends JavaFileObject> diagnostic : diagnostics.getDiagnostics()) {
+        compileRunService.sendMessages(principal.getName(), diagnostic.toString());
+      }
+
+      if (success) {
+        // set System.out to be a specific output stream in order to capture output of the
+        // program and send it back to the user
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        System.setOut(out);
+
+        compileRunService.sendMessages(principal.getName(), "Compiled!");
+        compileRunService.sendMessages(principal.getName(), "Running your program...");
+        runClass(tempFolder.toURI().toURL(), userProgram.getFileName(), principal);
+
+        // outputStream should now contain output of userProgram
+        outputStream.flush();
+        String result = outputStream.toString();
+        if (result.length() > 0) {
+          compileRunService.sendMessages(principal.getName(), result);
+        }
+      } else {
+        compileRunService.sendMessages(
+            principal.getName(), "There was an error compiling your program.");
+      }
+
+    } catch (IOException e) {
+      // IOException could be called by creating a temporary folder or writing to that folder.
+      // May need better error handling for this.
+      compileRunService.sendMessages(
+          principal.getName(), "There was an issue trying to run your program, please try again.");
+      System.out.println(e.getStackTrace());
+    }
+
+    if (tempFolder != null) {
+      tempFolder.delete();
+    }
+
+    // ensure System.out is reset
+    System.setOut(System.out);
+  }
+
+  // Given a user program, create a compilation task that will save the .class file to the given
+  // temp folder and output any compilation messages to diagnostics.
+  public CompilationTask getCompilationTask(
+      UserProgram userProgram, File tempFolder, DiagnosticCollector<JavaFileObject> diagnostics)
+      throws IOException {
+    JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
+
+    // set output of compilation to be a temporary folder
+    StandardJavaFileManager fileManager = compiler.getStandardFileManager(null, null, null);
+    fileManager.setLocation(StandardLocation.CLASS_OUTPUT, Arrays.asList(tempFolder));
+
+    // create file for user-provided code
+    JavaFileObject file =
+        new JavaSourceFromString(userProgram.getFileName(), userProgram.getCode());
+    Iterable<? extends JavaFileObject> compilationUnits = Arrays.asList(file);
+
+    // create compilation task
+    CompilationTask task =
+        compiler.getTask(null, fileManager, diagnostics, null, null, compilationUnits);
+    return task;
+  }
+
+  public void runClass(URL filePath, String className, Principal principal) {
```

Comment:
Possible convention: Do we want to javadoc comment all public methods?

Reply:
I think that is a good call, if we don't start with it we will never add them later. Will update this PR with javadoc comments + refactoring out logic

Reply:
This is something that the linter that I've used in the past (Checkstyle) can enforce.  Might be worth seeing if there is an equivalent feature in our linter (not necessarily as part of this commit).